### PR TITLE
Decouple polyscope raygen+cameras & move functionality to kaolin

### PR DESCRIPTION
**Pull request by orperel**
This MR is a preparation for the eventual headless mode of the playground. The motivation is to allow additional viewers where polyscope cannot be installed, as well as easier integration within other projects.

The `utils/kaolin_future` includes logic that will high probability be relocated to the next version of kaolin (v18.0).

1. `Playground` now assumes kaolin's [Cameras](https://github.com/NVIDIAGameWorks/kaolin/tree/master/examples/recipes/camera), instead of the gui-coupled `ViewParams` from polyscope. e.g. rendering is now invoked as `render(self, camera: Camera)`. kaolin and polyscope cameras can be converted back and forth, see `utils/kaolin_future/conversions.py`.
2. Playground rendering is offered in 2 ways: online (`render_pass()` - which is more interactive) and offline (`render()` - which calls all rendering passes needed for a single frame for high quality image), see `playground.py`.
3. **raygen**: functionality now uses Kaolin's raygen for pinhole rays. The raygen functionality for fisheye have been adapted to kaolin, see see `utils/kaolin_future/fisheye.py`.
4. **`VideoRecorder`** is extracted from `Playground` class to reduce the size of this monolith. All camera interpolations now use kaolin's new camera interpolation module, see `utils/kaolin_future/interpolated_cameras.py`.

For video recording, the gui have been slightly simplified by reducing the `use_dof` and `continuous` checkboxes to a combo box with the following options:

- mode (str): Determines how keyframes are interpolated:
  - 'linear_smooth' - shifts the camera in a linear path along the trajectory, using smoothstep interpolation (in main, this is _not use_dof and not continuous_)
  - 'linear_spline' - shifts the camera in a linear path along the trajectory, using catmull-rom splines (new mode)
  - 'continuous' - shifts the camera in a cyclic trajectory, determined by fitting a bspline (in main, this is _not use_dof and continuous_)
  - 'depth_of_field' - interpolates the depth of field over a static frame (first frame in trajectory) (in main, this is _use_dof_)

### CODE REVIEW

**Review by shumash — CHANGES_REQUESTED** [review]

File: playground/utils/kaolin_future/interpolated_cameras.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            cam4.fov(in_degrees=False)
+        )
+    elif cam1.lens_type == 'ortho' and cam2.lens_type == 'ortho':
+        intrinsics['fov_distance'] = \
+            (cam1.fov_distance(), cam2.fov_distance(), cam3.fov_distance(), cam4.fov_distance())
+
+    Xs = np.linspace(0.0, 1.0, frames_between_cameras)
+    t = Xs[timestep % frames_between_cameras]
+
+    eye = _catmull_rom(t, eye_1, eye_2, eye_3, eye_4)
```
Comment (nmwsharp) [inline playground/utils/kaolin_future/interpolated_cameras.py:196]: using some kind of lie algebra / quaternion interpolation rather than vector interpolation probably makes sense here as well

File: playground/utils/video_out.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                                                self.video_fps, (rgb.shape[2], rgb.shape[1]), True)
+                data = rgb[0].clip(0, 1).detach().cpu().numpy()
+                data = (data * 255).astype(np.uint8)
+                data = cv2.cvtColor(data, cv2.COLOR_BGR2RGB)
+                out_video.write(data)
+        finally:
+            self.renderer.use_depth_of_field = old_use_dof
+            self.renderer.depth_of_field.focus_z = old_focus_z
+
+    def render_linear_trajectory(self, interpolation_mode='polynomial'):
```
Comment (nmwsharp) [inline playground/utils/video_out.py:126]: the "linear" naming seems a little confusing here. What in what sense is polynomial interpolation "linear"? The differences of linear vs. continuous is not obvious to me from the names.

**Review by nmwsharp — APPROVED** [review]
Looks good to me! Comments from me are minor. The structural change decoupling the cameras from Polyscope makes sense. For the trajectory renderer, even if we could imagine some added features it doesn't have to be perfect here.

File: playground/utils/kaolin_future/conversions.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        height (int): Image plane height in pixels.
+        near (optional, float): near clipping plane, defines the min depth of the view frustrum.
+        far (optional, float): far clipping plane, define the max depth of the view frustrum.
+        device (optional, torch.device or str): the device on which camera parameters will be allocated. Default: cpu
+    Returns:
+        (kaolin.render.camera.Camera):
+            A kaolin camera object.
+    """
+    view_matrix = ps_camera.get_view_mat()
+    fov_y = ps_camera.get_fov_vertical_deg() * np.pi / 180.0    # to radians
```
Comment (zgojcic) [inline playground/utils/kaolin_future/conversions.py:52]: numpy also provides `np.deg2rad()` for convenience and then the comment is not needed

File: playground/utils/kaolin_future/transform.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        self._permutation = torch.eye(4, device=self.device, dtype=self.dtype)
+
+    def reset(self):
+        """ Restore object transform to unit scale, at the origin, with zero orientation. """
+        self._translation = torch.zeros(3, device=self.device, dtype=self.dtype)
+        self._rotation = torch.zeros(3, device=self.device, dtype=self.dtype)
+        self._scale = torch.ones(4, device=self.device, dtype=self.dtype)
+        self._permutation = torch.eye(4, device=self.device, dtype=self.dtype)
+
+    def translate(self, translation: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (orperel) [reply playground/utils/kaolin_future/transform.py:52]: @shumash 
The particular order of transformations comes from the object2world matrix (so called model matrix) convention used in graphics pipelines. This class was so far used in the context of gui-sliders and viewers, for representing geometric modifications like rotation and scale in place. e.g. the reference point is always the origin of the object in local coordinates. 

The reason its built this way is that for most user manipulations fixing the reference point is intuitive. Consider this common scenario in the playground:
* Armadillo inserted
* Move to (50, 100) so it appears above floor
* Rotate n degrees along y axis to align feet with floor
* Feet now intersect the ground -> move up 5 units
* Rotate 2 more degrees to properly align feet on the ground
...

For a more general use case for kaolin - we could of couse add a dynamic reference point, to stack transformations in any order
